### ISSUE TITLE
Add support for structs

### ISSUE THREAD

**Issue by Circuit8** [mceachen/exiftool-vendored.js#39]
Hey Matthew,

One thing I've noticed exiftool-vendored is missing is XMP struct tag support.
exiftool has the ability to [read and write struct information](https://www.sno.phy.queensu.ca/~phil/exiftool/struct.html#Fields).

If you want me to have a crack I can but my Typescript is still dire!

Cheers,
Josh

**Comment by mceachen** [comment]
This sounds like a great idea.

Is it as simple as adding `-struct -xmp:all`? Do you have an example image with XMP data I can play with?

**Comment by Circuit8** [comment]
So the command for writing I think would be something along the lines of this:

```bash
exiftool -RegistryID="[
  {RegItemId=Number1,RegOrgId=TestName1}, 
  {RegItemId=Number2,RegOrgId=TestName2},
  {RegItemId=Number3,RegOrgId=TestName3}
]" black.jpg
```

This sets the `RegistryID` field, with the `RegItemId` and `RegOrgId` fields. I believe that each struct field defines the permitted list of key names on the exiftool site, although probably best to just let exiftool handle that when writing and return any errors from it.

For reading we'd need something like `exiftool -struct image.jpg` which then returns the data in a parsable way I think.

I've attached an image with some struct data saved on it. Below is the snippet of how the XMP is stored in the image:

````xml
 <rdf:Description rdf:about=''
  xmlns:Iptc4xmpExt='http://iptc.org/std/Iptc4xmpExt/2008-02-29/'>
  <Iptc4xmpExt:RegistryId>
   <rdf:Bag>
    <rdf:li rdf:parseType='Resource'>
     <Iptc4xmpExt:RegItemId>Number1</Iptc4xmpExt:RegItemId>
     <Iptc4xmpExt:RegOrgId>TestName1</Iptc4xmpExt:RegOrgId>
    </rdf:li>
    <rdf:li rdf:parseType='Resource'>
     <Iptc4xmpExt:RegItemId>Number2</Iptc4xmpExt:RegItemId>
     <Iptc4xmpExt:RegOrgId>TestName2</Iptc4xmpExt:RegOrgId>
    </rdf:li>
    <rdf:li rdf:parseType='Resource'>
     <Iptc4xmpExt:RegItemId>Number3</Iptc4xmpExt:RegItemId>
     <Iptc4xmpExt:RegOrgId>TestName3</Iptc4xmpExt:RegOrgId>
    </rdf:li>
   </rdf:Bag>
  </Iptc4xmpExt:RegistryId>
 </rdf:Description>
````
![black](https://user-images.githubusercontent.com/6171759/50146712-537f6880-02ac-11e9-8c6b-a7cd3c21e677.jpg)


Let me know if you need anything else!

**Comment by mceachen** [comment]
> I don't understand why this is a new issue though, wasn't there always risk of group namespace collision?

You're right, and in mktags I actually had to remove tags that collide, but those tags seemed to have the same semantic value.

Think `FlashFired`: I'd be surprised if the messing of that tag changed based on what group you found it in. 

Something generic like `History`, though, seems like it should be quarantined to its own group to disambiguate.

Does that make sense?

**Comment by Circuit8** [comment]
Yeah I think so.

One idea is we could use the call to exiftool to get the grouped data like you suggested, but then rather than returning that to the user, we collate it all into one object. If there are any collisions with the same key in different namespaces we prepend the namespace into the results, e.g:

``` JSON
{
  "Non-Clashing-Field": 'some stuff',
  "IPTC:History": 'history from iptc',
  "xmp-dc:History": 'history from xmp',
}
```

The benefit of this is that its the same format as what exiftool expects if you're writing to a specific subgroup.

What do you reckon?

**Comment by mceachen** [comment]
I think I've gotten into trouble when I've done this in the past with APIs in an effort to be "nice".

What happens if the tag is colliding in one file and not in another? You'd have to look for `History` in two places.

I like the idea that the colon-separated tags are what ExifTool expects, so we aren't inventing Yet Another API.

This would make the tags objects larger, but what do you think about doing *both*: put all non-clashing fields into the groupless top level, *and* prefix every tag with their group name?

**Comment by Circuit8** [comment]
Actually the more I think about it the more I like your original solution `ExifTool.readWithGroups()`, just group everything by its namespace. Nice and clean, and gives you everything you need to know, and doesn't change if there's collisions or not, the field will just appear in more than one category.

**Comment by Circuit8** [comment]
Hey I've found some time to finally get my head around Typescript and have a go at this. Hopefully I'll have a PR for you in the next few days

**Comment by mceachen** [comment]
Excellent. If you get stuck, post the PR as a work-in-progress and I can help you finish it.

**Comment by Circuit8** [comment]
I've managed to get a PR together https://github.com/mceachen/exiftool-vendored.js/pull/43. Let me know what you think.

BTW I'm starting to see why Typescript could be useful, I was quite anti TS before but after doing some reading I'm beginning to see the benefits. Are you using it in your other projects and do you feel the safety gains are worth the extra overhead?

**Comment by mceachen** [comment]
I've lived with very large C, C++, java, ruby, python, and scala projects, and find that the compiler is definitely your friend in all cases. The large ruby and python projects I've worked on were extremely hard to make confident changes in. I wouldn't develop in javascript directly if I had the choice.

**Comment by mceachen** [comment]
Your PR just adds `-struct` to all read tasks. Did you decide that `readWithGroups` was not worth it?

**Comment by mceachen** [comment]
Thanks for the PR! Just released v8.0.0.
